### PR TITLE
Pass the source account into the host and allow calling into the token wrapper host functions

### PR DESCRIPTION
**Pull request by sisuresh**
# Description

Resolves https://github.com/stellar/stellar-core/issues/3546 and https://github.com/stellar/stellar-core/issues/3547.

# Checklist
- [ ] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [ ] Rebased on top of master (no merge commits)
- [ ] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [ ] Compiles
- [ ] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/transactions/test/InvokeHostFunctionTests.cpp [hunk 1]
```diff
@@ -65,10 +65,39 @@ makeSymbol(std::string const& str)
...
+    REQUIRE(tx->checkValid(ltx, 0, 0, 0));
+    REQUIRE(tx->apply(app, ltx, txm) == expectSuccess);
+    ltx.commit();
+
+    auto contractCodeObj =
+        makeContract(contract.vec.begin(), contract.vec.end());
+    // verify contract code is correct
+    LedgerTxn ltx2(app.getLedgerTxnRoot());
+    auto ltxe2 = loadContractData(ltx2, contractID, wasmKey);
+    REQUIRE((bool)ltxe2 == expectEntry);
```
Comment (dmkozh) [inline src/transactions/test/InvokeHostFunctionTests.cpp:88]: nit: Please use `static_cast` - c-style casts should normally be avoided in c++.

File: src/transactions/test/InvokeHostFunctionTests.cpp [hunk 2]
```diff
@@ -99,45 +126,79 @@ createContract(Application& app, Bytes const& contract, uint256 const& salt,
...
-    hasher.add(contract);
-
-    auto sig = SignatureUtils::sign(key, hasher.finish()).signature;
-    return createContract(app, contract, salt, key.getPublicKey(), sig, true,
-                          true);
+    switch (fn)
+    {
+    case HostFunction::HOST_FN_CREATE_CONTRACT_WITH_ED25519:
+    {
+        // uint256 salt = sha256("salt");
```
Comment (dmkozh) [inline src/transactions/test/InvokeHostFunctionTests.cpp:179]: Please remove this.

**Review by dmkozh — COMMENTED** [review]
LGTM, modulo a couple nitpicks

File: src/transactions/test/InvokeHostFunctionTests.cpp (hunk 1 above)
Comment (sisuresh) [reply src/transactions/test/InvokeHostFunctionTests.cpp]: Done

File: src/transactions/test/InvokeHostFunctionTests.cpp (hunk 2 above)
Comment (sisuresh) [reply src/transactions/test/InvokeHostFunctionTests.cpp]: Done

**Comment by sisuresh** [comment]
Marked as draft so we can get https://github.com/stellar/stellar-core/issues/3547 in as well.

**Review by graydon — APPROVED** [review]
One nit I notice, everything else looks reasonable from what I understand about the goal here. I haven't been following the development of these APIs very closely though, so might want second eyes if you're looking for design feedback!

File: src/rust/src/contract.rs [hunk 3]
```diff
@@ -194,26 +194,53 @@ pub(crate) fn invoke_host_function(
...
+    host.set_source_account(source_account);
 
     match hf {
-        HostFunction::Call => {
-            info!(target: TX, "Invoking host function 'Call'");
+        HostFunction::InvokeContract => {
+            info!(target: TX, "Invoking host function 'InvokeContract'");
             host.invoke_function(hf, args)?;
         }
-        HostFunction::CreateContract => {
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/rust/src/contract.rs:215]: Good point. Fixed.

**Comment by graydon** [comment]
r+ 060a6bb

**Pull request merged**